test/ha/activator_test.go

```diff
@@ -0,0 +1,101 @@
+)
+
+const (
+	activatorDeploymentName = "activator"
+	activatorLabel          = "app=activator"
+)
+
+// Activator is managed by the activator HPA and does not have leader election enabled.
+// The test ensures that stopping one of the activator pods doesn't affect user applications.
+func TestActivatorHA(t *testing.T) {
+	clients := e2e.Setup(t)
+
+	names, resources := createPizzaPlanetService(t,
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.MinScaleAnnotationKey:  "1",  //make sure we don't scale to zero during the test
+			autoscaling.TargetBurstCapacityKey: "-1", // make sure all requests go through the activator
+		}),
+	)
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	namesScaleToZero, resourcesScaleToZero := createPizzaPlanetService(t,
+		rtesting.WithConfigAnnotations(map[string]string{
+			autoscaling.WindowAnnotationKey:    autoscaling.WindowMin.String(), //make sure we scale to zero quickly
+			autoscaling.TargetBurstCapacityKey: "-1",
+		}),
+	)
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, namesScaleToZero) })
+	defer test.TearDown(clients, namesScaleToZero)
+
+	if err := e2e.WaitForScaleToZero(t, revisionresourcenames.Deployment(resourcesScaleToZero.Revision), clients); err != nil {
+		t.Fatalf("Failed to scale to zero: %v", err)
+	}
+
+	prober := test.RunRouteProber(log.Printf, clients, resources.Service.Status.URL.URL())
+	defer test.AssertProberDefault(t, prober)
+
+	pods, err := clients.KubeClient.Kube.CoreV1().Pods(servingNamespace).List(metav1.ListOptions{
+		LabelSelector: activatorLabel,
+	})
+	if err != nil {
+		t.Fatalf("Failed to get activator pods: %v", err)
+	}
+	if len(pods.Items) != haReplicas {
+		t.Fatalf("The test requires %d activator pods", haReplicas)
+	}
+	activatorPod := pods.Items[0].Name // stop the first activator pod
+
+	clients.KubeClient.Kube.CoreV1().Pods(servingNamespace).Delete(activatorPod, &metav1.DeleteOptions{})
+
+	if err := waitForPodDeleted(t, clients, activatorPod); err != nil {
+		t.Fatalf("Did not observe %s to actually be deleted: %v", activatorPod, err)
+	}
+
+	if err := waitForDeploymentScale(clients, activatorDeploymentName, haReplicas); err != nil {
+		t.Fatalf("Deployment %s failed to scale up: %v", activatorDeploymentName, err)
+	}
+
+	activatorPod = pods.Items[1].Name // now stop the other activator pod
+
+	clients.KubeClient.Kube.CoreV1().Pods(servingNamespace).Delete(activatorPod, &metav1.DeleteOptions{})
+	if err := waitForPodDeleted(t, clients, activatorPod); err != nil {
+		t.Fatalf("Did not observe %s to actually be deleted: %v", activatorPod, err)
+	}
+
+	if err := waitForDeploymentScale(clients, activatorDeploymentName, haReplicas); err != nil {
+		t.Fatalf("Deployment %s failed to scale up: %v", activatorDeploymentName, err)
+	}
+
+	assertServiceWorks(t, clients, namesScaleToZero, resourcesScaleToZero.Service.Status.URL.URL(), test.PizzaPlanetText1)
```

Comment:
So is this testing that after the two activators are back up, we can scale from 0 successfully? 🤔

Reply:
I could move it right after stopping the first activator and before I wait for it to be deleted?